server/app/controllers/applicant/ApplicantProgramsController.java

```diff
@@ -162,6 +168,26 @@ public CompletionStage<Result> view(Request request, long applicantId, long prog
             });
   }
 
+  // This controller method disambiguates between two routes:
+  //
+  // - /programs/<program-id>
+  // - /programs/<program-slug>
+  //
+  // Because the second use is public, this controller is not annotated as @Secure. For the first
+  // use, the delegated-to method *is* annotated as such.
+  public CompletionStage<Result> view(Request request, String programParam) {
+    if (StringUtils.isNumeric(programParam)) {
+      // The path parameter specifies a program by (numeric) id.
+
+      // The route for this action should only be computed if the applicant ID is available in the
+      // session.
+      long applicantId = getApplicantId(request).orElseThrow();
```

Comment:
**issue (blocking):** throwing here will produce a 500 (server error), indicating a problem in the server code. If this throws though it's the result of a bad request -- an HTTP client without a session attempting to access a route they can't be authenticated for.

One option would be to return a 4XX level status code instead, which will at least help our monitoring be more accurate.

Something else to consider is when this may occur -- e.g. if someone bookmarks the URL for an application they're working on and then their session expires. Is there a way to redirect them to the identity provider instead?

Reply:
This last point is an excellent one that had not occurred to me. I've made the change here and will need to go back and make the corresponding change in #5946 before merging.